core/src/main/java/com/radixdlt/sync/SyncResponseHandler.java

```diff
@@ -0,0 +1,172 @@
+import com.radixdlt.consensus.LedgerProof;
+import com.radixdlt.environment.EventDispatcher;
+import com.radixdlt.ledger.*;
+import com.radixdlt.p2p.NodeId;
+import com.radixdlt.statecomputer.commit.InvalidCommitRequestError;
+import com.radixdlt.sync.SyncState.SyncingState;
+import com.radixdlt.sync.messages.local.*;
+import com.radixdlt.sync.messages.remote.*;
+import com.radixdlt.sync.validation.RemoteSyncResponseSignaturesVerifier;
+import com.radixdlt.sync.validation.RemoteSyncResponseValidatorSetVerifier;
+
+/** An implementation delegate of {@link LocalSyncService}, for handling the sync responses. */
+public final class SyncResponseHandler {
+
+  private final RemoteSyncResponseValidatorSetVerifier validatorSetVerifier;
+  private final RemoteSyncResponseSignaturesVerifier signaturesVerifier;
+  private final EventDispatcher<LedgerExtension> syncedLedgerExtensionDispatcher;
+
+  @Inject
+  public SyncResponseHandler(
+      RemoteSyncResponseValidatorSetVerifier validatorSetVerifier,
+      RemoteSyncResponseSignaturesVerifier signaturesVerifier,
+      EventDispatcher<LedgerExtension> syncedLedgerExtensionDispatcher) {
+    this.validatorSetVerifier = validatorSetVerifier;
+    this.signaturesVerifier = signaturesVerifier;
+    this.syncedLedgerExtensionDispatcher = syncedLedgerExtensionDispatcher;
+  }
+
+  /**
+   * Commits the transactions from the given sync response to the ledger. Throws a {@link
+   * InvalidSyncResponseException} in case of any validation failure (including current syncing
+   * state, ledger state, validator signatures, etc. - see the sealed exception subclasses for
+   * details).
+   */
+  public void handle(SyncingState currentState, NodeId sender, SyncResponse syncResponse) {
+    final var dto = syncResponse.getLedgerExtension();
+    final var start = LedgerProof.fromDto(dto.getStart());
+    final var end = LedgerProof.fromDto(dto.getEnd());
+    final var ledgerExtension = LedgerExtension.create(dto.getTransactions(), end);
+    this.checkMatchesPendingRequest(currentState, sender, start);
+    this.checkValidTransactionCount(start, ledgerExtension);
+    this.checkConsensusProof(end);
+    this.commit(ledgerExtension);
+  }
+
+  private void checkMatchesPendingRequest(
+      SyncingState currentState, NodeId sender, LedgerProof ledgerExtensionStart) {
+
+    final var optPendingRequestPeer =
+        currentState.getPendingRequest().map(SyncState.PendingRequest::getPeer);
+    if (optPendingRequestPeer.isEmpty()) {
+      throw new InvalidSyncResponseException.NoSyncRequestPending();
+    }
+
+    final var pendingRequestPeer = optPendingRequestPeer.get();
+    if (!pendingRequestPeer.equals(sender)) {
+      throw new InvalidSyncResponseException.SyncRequestSenderMismatch();
+    }
+
+    final var currentHeader = currentState.getCurrentHeader();
+    if (!ledgerExtensionStart.equals(currentHeader)) {
+      throw new InvalidSyncResponseException.LedgerExtensionStartMismatch();
```

Comment:
I believe this particular check is more specific than present - we used to accept through requests with a start not matching our current state. Out of interest, what motivated this addition? I guess you felt that this was missing before and needed adding? (It's not wrong, I'm just curious)

In fact I don't think the start proof actually matters (beyond saying "this is where you requested and where I start from").

There are probably some cases where ignoring this start makes things more efficient (eg BFT sync in the mean time) - but, that said, I can't think of any which would be too hurt by this check... 

I think on balance it's better to keep the check in, but interested to hear your thoughts.

Reply:
Yup, I added this on purpose and here is my reasoning:

The whole "ledger sync" feature looks like an RPC use-case: there is a Request, there is a Response.
In fact, we have like ~80% impl of a poor-man's RPC infra here: there is tracking of pending exchanges (request-response pairs), there is rejection of unsolicited responses, there is even "request ID" (but used only for timeout-tracking).

A true RPC infra would actually transmit that request ID to the peer (and expect it in the matching response). Alternatively, it could expect a `hash(request)` in the matching response. Ooooor, alternatively (and a bit wastefully), it could expect a verbatim `request` in the matching response.

So I thought "well, we have the `request` already echoed back in the response, let's expect it to match".
There is no reason (for a well-behaving peer) to return any older start header - if it happens, then most likely our state changed in the meantime and we are no longer expecting this particular sync response.
I prefer to match (or fail-fast) and later in the code have a guarantee that "yes, this is the response to exactly my request, it arrived in time, and I am still in the perfect state to process it as-is".

> There are probably some cases where ignoring this start makes things more efficient

I agree, totally - but I don't think it's worth optimizing for "making maximum use of corner-cased/timed-out/unsolicited sync responses".

And an argument from a different angle:
Please note that the same is true about "ignoring the sender of the response" (I mean: if some other peer just sends you a nice, useful, correctly-signed ledger extension out of nowhere, then why not apply it? Yet, we always match on the sender [even before this PR]).

Reply:
Sounds good to me